src/core/radio/trel_interface.cpp

```diff
@@ -34,101 +34,421 @@
+    string.AppendHexBytes(Get<Mac::Mac>().GetExtAddress().m8, sizeof(Mac::ExtAddress));
+
+    *txtPtr++ = static_cast<uint8_t>(string.GetLength());
+    memcpy(txtPtr, string.AsCString(), string.GetLength());
+    txtPtr += string.GetLength();
+
+    string.Clear();
+    string.Append("%s=", kTxtRecordExtPanIdKey);
+    string.AppendHexBytes(Get<Mac::Mac>().GetExtendedPanId().m8, sizeof(Mac::ExtendedPanId));
+
+    *txtPtr++ = static_cast<uint8_t>(string.GetLength());
+    memcpy(txtPtr, string.AsCString(), string.GetLength());
+    txtPtr += string.GetLength();
+
+    OT_ASSERT(txtPtr == OT_ARRAY_END(txtData));
+
+    otLogInfoMac("Trel: Registering DNS-SD service: port:%u, txt:\"%s=%s, %s=%s\"", mUdpPort, kTxtRecordExtAddressKey,
+                 Get<Mac::Mac>().GetExtAddress().ToString().AsCString(), kTxtRecordExtPanIdKey,
+                 Get<Mac::Mac>().GetExtendedPanId().ToString().AsCString());
+
+    otPlatTrelRegisterService(&GetInstance(), mUdpPort, txtData, sizeof(txtData));
+
+exit:
+    return;
+}
+
+extern "C" void otPlatTrelHandleDiscoveredPeerInfo(otInstance *aInstance, const otPlatTrelPeerInfo *aInfo)
+{
+    Instance &instance = AsCoreType(aInstance);
+
+    VerifyOrExit(instance.IsInitialized());
+    instance.Get<Interface>().HandleDiscoveredPeerInfo(*static_cast<const Interface::Peer::Info *>(aInfo));
+
+exit:
+    return;
+}
+
+void Interface::HandleDiscoveredPeerInfo(const Peer::Info &aInfo)
+{
+    Peer *             entry;
+    Mac::ExtAddress    extAddress;
+    Mac::ExtendedPanId extPanId;
+    bool               isNew = false;
+
+    VerifyOrExit(mInitialized && mEnabled);
+
+    SuccessOrExit(ParsePeerInfoTxtData(aInfo, extAddress, extPanId));
+
+    VerifyOrExit(extAddress != Get<Mac::Mac>().GetExtAddress());
+
+    if (aInfo.IsRemoved())
+    {
+        entry = mPeerTable.FindMatching(extAddress);
+        VerifyOrExit(entry != nullptr);
+        RemovePeerEntry(*entry);
+        ExitNow();
+    }
+
+    // It is a new entry or an update to an existing entry. First
+    // check whether we have an existing entry that matches the same
+    // socket address, and remove it if it is associated with a
+    // different Extended MAC address. This ensures that we do not
+    // keep stale entries in the peer table.
+
+    entry = mPeerTable.FindMatching(aInfo.GetSockAddr());
+
+    if ((entry != nullptr) && !entry->Matches(extAddress))
+    {
+        RemovePeerEntry(*entry);
+        entry = nullptr;
+    }
+
+    if (entry == nullptr)
+    {
+        entry = mPeerTable.FindMatching(extAddress);
+    }
+
+    if (entry == nullptr)
+    {
+        entry = GetNewPeerEntry();
+        VerifyOrExit(entry != nullptr);
+
+        entry->SetExtAddress(extAddress);
+        isNew = true;
+    }
+
+    if (!isNew)
+    {
+        VerifyOrExit((entry->GetExtPanId() != extPanId) || (entry->GetSockAddr() != aInfo.GetSockAddr()));
+    }
+
+    entry->SetExtPanId(extPanId);
+    entry->SetSockAddr(aInfo.GetSockAddr());
```

Comment:
TREL does not define explicitly which network interface platform should be browsing TREL services. 
A straightforward platform implementation would browse on all interfaces, which is usually default on mDNS responders. 

So, it's possible that one TREL peer can be discovered multiple times on different interfaces. 

So, the current logic will constantly replace the socket address whenever a TREL service is discovered on any network interface. 

So, does this imply that platform TREL DNS-SD should only be browsing on one network interface?
Or are we saying that it's fine since the socket address should work no matter which interface it's discovered on..

Reply:
Yes. I agree. All DNS-SD activities should be limited to the network interface(s) on which we want to run TREL. I would say it is probably the same for the socket(s) we use to send and receive (we may want to bind them to certain interface).

I think it make sense to leave it up to platform/project to determine which interface(s) it wants to use for TREL. 

If there are multiple interfaces, then we may find the same peer on different interface and potentially with different addresses. In this case, I would say we can again leave it to platform to select a preferred address/interface from the list of addresses (somewhat the same way that if we discover multiple addresses associated with the same host/service (on the same interface), we try to pick the one with the larger scope).

Reply:
SGTM to let the platform choose the netif.